bpf/Makefile

```diff
@@ -19,46 +19,67 @@ $(BPF_SIMPLE): %.o: %.ll
 	@$(ECHO_CC)
 	$(QUIET) ${LLC} ${LLC_FLAGS} -filetype=obj -o $@ $(patsubst %.o,%.ll,$@)
 
+# Hack to get make to replace : with a space
+space :=
+space += 
+
+# The following option combinations are compile tested
 LB_OPTIONS = \
 	-DSKIP_DEBUG \
-	-DLB_L3 \
-	-DLB_L4 \
-	"-DLB_L3 -DLB_L4" \
-	"-DDISABLE_IPV4 -DLB_L3 -DLB_L4" \
-	"-DDISABLE_IPV6 -DLB_L3 -DLB_L4" \
-	"-DDISABLE_IPV4 -DDISABLE_IPV6 -DLB_L3 -DLB_L4"
-
+	-DENABLE_IPV4:-DLB_L3 \
+	-DENABLE_IPV4:-DLB_L4 \
+	-DENABLE_IPV4:-DLB_L3:-DLB_L4 \
+	-DENABLE_IPV6:-DLB_L3 \
+	-DENABLE_IPV6:-DLB_L4 \
+	-DENABLE_IPV6:-DLB_L3:-DLB_L4 \
+	-DENABLE_IPV4:-DENABLE_IPV6:-DLB_L3 \
+	-DENABLE_IPV4:-DENABLE_IPV6:-DLB_L4 \
+	-DENABLE_IPV4:-DENABLE_IPV6:-DLB_L3:-DLB_L4
+
+# These options are intended to max out the BPF program complexity. it is load
+# tested as well.
+MAX_LB_OPTIONS = -DENABLE_IPV4 -DENABLE_IPV6 -DLB_L3 -DLB_L4
 
 bpf_lb.ll: bpf_lb.c $(LIB)
 	$(QUIET) set -e; \
 	$(foreach OPTS,$(LB_OPTIONS), \
-		$(ECHO_CC) " [$(OPTS)]"; \
-		${CLANG} ${OPTS} ${CLANG_FLAGS} -c $< -o $@; \
+		$(ECHO_CC) " [$(subst :,$(space),$(OPTS))]"; \
+		${CLANG} $(subst :,$(space),$(OPTS)) ${CLANG_FLAGS} -c $< -o $@; \
 		${LLC} ${LLC_FLAGS} -o /dev/null $@; )
 	@$(ECHO_CC)
-	$(QUIET) ${CLANG} ${CLANG_FLAGS} -c $< -o $@
+	$(QUIET) ${CLANG} ${MAX_LB_OPTIONS} ${CLANG_FLAGS} -c $< -o $@
 
 bpf_lb.o: bpf_lb.ll
 	@$(ECHO_CC)
 	$(QUIET) ${LLC} ${LLC_FLAGS} -filetype=obj -o $@ $(patsubst %.o,%.ll,$@)
 
+# The following option combinations are compile tested
 LXC_OPTIONS = \
 	 -DSKIP_DEBUG \
 	 -DHAVE_LPM_MAP_TYPE \
 	 -DHAVE_LRU_MAP_TYPE \
-	 -DDISABLE_IPV4 \
-	 -DDISABLE_IPV6 \
-	 "-DDISABLE_IPV6 -DDISABLE_IPV4" \
-	 "-DDISABLE_IPV4 -DDISABLE_NAT46"
+	 -DENABLE_IPV4 \
+	 -DENABLE_IPV4:-DHAVE_LPM_MAP_TYPE \
+	 -DENABLE_IPV4:-DHAVE_LPM_MAP_TYPE:-DHAVE_LRU_MAP_TYPE \
+	 -DENABLE_IPV6 \
+	 -DENABLE_IPV6:-DHAVE_LPM_MAP_TYPE \
+	 -DENABLE_IPV6:-DHAVE_LPM_MAP_TYPE:-DHAVE_LRU_MAP_TYPE \
+	 -DENABLE_IPV6:-DENABLE_IPV4 \
+	 -DENABLE_IPV4:-DENABLE_IPV6:-DHAVE_LPM_MAP_TYPE:-DHAVE_LRU_MAP_TYPE \
+	 -DENABLE_IPV4:-DENABLE_NAT46
+
+# These options are intended to max out the BPF program complexity. it is load
+# tested as well.
+MAX_LXC_OPTIONS = "-DENABLE_IPV4 -DENABLE_IPV6"
 
 bpf_lxc.ll: bpf_lxc.c $(LIB)
 	$(QUIET) set -e; \
 	$(foreach OPTS,$(LXC_OPTIONS), \
-		$(ECHO_CC) " [$(OPTS)]"; \
-		${CLANG} ${OPTS} ${CLANG_FLAGS} -c $< -o $@; \
+		$(ECHO_CC) " [$(subst :,$(space),$(OPTS))]"; \
+		${CLANG} $(subst :,$(space),$(OPTS)) ${CLANG_FLAGS} -c $< -o $@; \
 		${LLC} ${LLC_FLAGS} -o /dev/null $@; )
 	@$(ECHO_CC)
```

Comment:
Might be nice to get the annotations here like we do for the compile options, ie `@$(ECHO_CC) " [${MAX_LXC_OPTONS}]"`

Reply:
The entire command line is printed:
```
clang "-DENABLE_IPV4 -DENABLE_IPV6" -I/home/vagrant/go/src/github.com/cilium/cilium/bpf/include -I/home/vagrant/go/src/github.com/cilium/cilium/bpf -D__NR_CPUS__=2 -O2 -g -target bpf -emit-llvm -Wall -Werror -Wno-address-of-packed-member -Wno-unknown-warning-option -c bpf_lxc.c -o bpf_lxc.ll
```

I don't actually understand the logic of the existing annotations. It seems that all annotations are printed first and then all compilations happen. I guess it's useful to see what combinations are tested but it doesn't actually allow to correlate a compile error to a combination.

Reply:
Here's how it works for me (I shortcut all the make parameters with aliases though):

```
$ git di                                                                                                                                     
diff --git a/bpf/lib/common.h b/bpf/lib/common.h                                                                                                                                         
index 5a31d0dc784d..659ddbe02096 100644                                                                                                                                                  
--- a/bpf/lib/common.h                                                                                                                                                                   
+++ b/bpf/lib/common.h                                                                                                                                                                   
@@ -60,6 +60,10 @@                                                                                                                                                                       
 #define CILIUM_CALL_IPV6_TO_LXC                        12                                                                                                                               
 #define CILIUM_CALL_SIZE                       13                                                                                                                                       
                                                                                                                                                                                         
+#if defined(ENABLE_IPV4) && defined(ENABLE_IPV6)                                                                                                                                        
+invalid-c                                                                                                                                                                               
+#endif                                                                                                                                                                                  
+                                                                                                                                                                                        
 typedef __u64 mac_t;                                                                                                                                                                    
                                                                                                                                                                                         
 union v6addr {                                                                                                                                                                          
$ make --quiet V=0 -j1 -C bpf                                                                                                                
  CC    bpf/bpf_lxc.ll  [-DSKIP_DEBUG]                                                                                                                                                   
  CC    bpf/bpf_lxc.ll  [-DHAVE_LPM_MAP_TYPE]                                                                                                                                            
  CC    bpf/bpf_lxc.ll  [-DHAVE_LRU_MAP_TYPE]                                                                                                                                            
  CC    bpf/bpf_lxc.ll  [-DENABLE_IPV4]                                                                                                                                                  
  CC    bpf/bpf_lxc.ll  [-DENABLE_IPV4 -DHAVE_LPM_MAP_TYPE]                                                                                                                              
  CC    bpf/bpf_lxc.ll  [-DENABLE_IPV4 -DHAVE_LPM_MAP_TYPE -DHAVE_LRU_MAP_TYPE]                                                                                                          
  CC    bpf/bpf_lxc.ll  [-DENABLE_IPV6]                                                                                                                                                  
  CC    bpf/bpf_lxc.ll  [-DENABLE_IPV6 -DHAVE_LPM_MAP_TYPE]
  CC    bpf/bpf_lxc.ll  [-DENABLE_IPV6 -DHAVE_LPM_MAP_TYPE -DHAVE_LRU_MAP_TYPE]
  CC    bpf/bpf_lxc.ll  [-DENABLE_IPV6 -DENABLE_IPV4]
In file included from bpf_lxc.c:31:
/home/joe/work/src/github.com/cilium/cilium/bpf/lib/common.h:64:1: error: unknown type name 'invalid'                                                                                   
invalid-c
^
/home/joe/work/src/github.com/cilium/cilium/bpf/lib/common.h:64:8: error: expected identifier or '('                                                                                    
invalid-c
       ^
/home/joe/work/src/github.com/cilium/cilium/bpf/lib/common.h:178:2: error: unknown type name 'mac_t'                                                                                    
        mac_t           mac;
        ^
/home/joe/work/src/github.com/cilium/cilium/bpf/lib/common.h:179:2: error: unknown type name 'mac_t'                                                                                    
        mac_t           node_mac;
        ^
/home/joe/work/src/github.com/cilium/cilium/bpf/lib/common.h:526:2: error: unknown type name 'mac_t'                                                                                    
        mac_t node_mac;
        ^
...
```